### ISSUE TITLE
Carriage return (\r) dropped by `XMLStreamWriter` implementation

### ISSUE THREAD

**Issue by conleym** [FasterXML/aalto-xml#35]
I'm using StAX to do some XML processing, reading events and then writing them, sometimes with some changes, etc., as you do, and I've noticed something odd: aalto, given XML containing `&#13;` in a text node will, at the end of this process, produce text with neither the CR nor a newline. This differs from other StAX implementations and appears to be a bug (the handling of this particular case is different across all implementations I know of, but aalto's seems to be the most-wrong).

This program illustrates:

```
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.nio.charset.StandardCharsets;

import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.stream.XMLEventFactory;
import javax.xml.stream.XMLEventReader;
import javax.xml.stream.XMLEventWriter;
import javax.xml.stream.XMLInputFactory;
import javax.xml.stream.XMLOutputFactory;
import javax.xml.stream.events.XMLEvent;
import javax.xml.xpath.XPath;
import javax.xml.xpath.XPathConstants;
import javax.xml.xpath.XPathFactory;

import org.w3c.dom.Document;


public final class Minimal {
    /** Enumeration of supported StAX implementations. */
    public enum StAXImplementation {
        AALTO("com.fasterxml.aalto.stax.InputFactoryImpl",
                "com.fasterxml.aalto.stax.OutputFactoryImpl",
                "com.fasterxml.aalto.stax.EventFactoryImpl"),
        /** JDK built in implementation, based on Xerces. */
        JDK("com.sun.xml.internal.stream.XMLInputFactoryImpl",
                "com.sun.xml.internal.stream.XMLOutputFactoryImpl",
                "com.sun.xml.internal.stream.events.XMLEventFactoryImpl"),
        WOODSTOX("com.ctc.wstx.stax.WstxInputFactory",
                "com.ctc.wstx.stax.WstxOutputFactory",
                "com.ctc.wstx.stax.WstxEventFactory"),
        XERCES(JDK.inputFactory,
                JDK.outputFactory,
                "org.apache.xerces.stax.XMLEventFactoryImpl");

        final String inputFactory;
        final String outputFactory;
        final String eventFactory;

        private StAXImplementation(final String inputFactory,
                final String outputFactory, final String eventFactory) {
            this.inputFactory = inputFactory;
            this.outputFactory = outputFactory;
            this.eventFactory = eventFactory;
        }

        /**
         * Tell the JDK to use this StAXImplementation.
         */
        public void use() {
            System.setProperty("javax.xml.stream.XMLInputFactory",
                    inputFactory);
            System.setProperty("javax.xml.stream.XMLOutputFactory",
                    outputFactory);
            System.setProperty("javax.xml.stream.XMLEventFactory",
                    eventFactory);
        }
    }


    static final String CR = new String(Character.toChars(32));
    static final String TEXT="a&#13;a";
    static final String EXPANDED_TEXT = "a" + CR + "a";
    static final String NEWLINE_TEXT = "a\na";
    static final String XML = "<x>" + TEXT + "</x>";
    static final byte[] XML_BYTES = XML.getBytes(
            StandardCharsets.UTF_8);

    /** Run some tests for each StAX implementation. */
    public static void main(final String[] args) throws Exception {
        for (final StAXImplementation impl : StAXImplementation.values()) {
            runTest(impl);
            System.out.println();
        }
    }

    private static void runTest(final StAXImplementation impl)
            throws Exception {
        impl.use();
        System.out.println("************** Trying " + impl + " **************");
        final XMLInputFactory inputFactory = XMLInputFactory.newFactory();
        final XMLOutputFactory outputFactory =
                XMLOutputFactory.newFactory();
        final ByteArrayOutputStream baos = new ByteArrayOutputStream();

        System.out.println("-------------- Factory classes:");
        printName("XMLInputFactory", inputFactory);
        printName("XMLOutputFactory", outputFactory);
        printName("XMLEventFactory", XMLEventFactory.newFactory());

        System.out.println("-------------- StAX results:");
        final XMLEventReader r = inputFactory.createXMLEventReader(
                new ByteArrayInputStream(XML_BYTES));
        final XMLEventWriter w = outputFactory.createXMLEventWriter(baos);
        final StringBuilder buffer = new StringBuilder();
        while (r.hasNext()) {
            final XMLEvent e = r.nextEvent();
            if (e.isStartDocument()) {
                // Avoid the XML declaration. Not present in the input.
                continue;
            }
            if (e.isCharacters()) {
                buffer.append(e.asCharacters().getData());
            } else {
                if (buffer.length() > 0) {
                    testText(buffer.toString());
                    buffer.setLength(0);
                }
            }
            w.add(e);
        }
        r.close();
        w.flush();
        w.close();
        final byte[] resultBytes = baos.toByteArray();
        System.out.println("StAX XML: [" + new String(resultBytes,
                StandardCharsets.UTF_8) + "]");
        testDOM(resultBytes);
    }

    private static void printName(final String name, final Object obj) {
        System.out.println(name + "=" + obj.getClass().getName());
    }

    private static void testText(final String text) {
        System.out.println("Buffered text: [" + text + "]");
        System.out.println("Code point at index 1: " +
                Character.codePointAt(text, 1));
        System.out.println("Buffered text equals input text? " +
                TEXT.equals(text));
        System.out.println(
                "Buffered text equals expanded text? " +
                        EXPANDED_TEXT.equals(text));
        System.out.println("Buffered text has \\n for \\r? " +
                NEWLINE_TEXT.equals(text));
    }

    private static void testDOM(final byte[] resultBytes) throws Exception {
        final DocumentBuilderFactory dbf =
                DocumentBuilderFactory.newInstance();
        dbf.setNamespaceAware(true);
        dbf.setExpandEntityReferences(false);
        final Document doc = dbf.newDocumentBuilder().parse(
                new ByteArrayInputStream(resultBytes));
        final XPath xpath = XPathFactory.newInstance().newXPath();
        final String domText = (String)xpath.evaluate("/x/text()", doc,
                XPathConstants.STRING);
        System.out.println("============== DOM results:");
        testText(domText);
    }
}
```

It's pretty straightforward to run if you have all the implementations in your classpath. I use the following pom:

```
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>
  <groupId>foobar</groupId>
  <artifactId>stax-stuff</artifactId>
  <version>0.0.1-SNAPSHOT</version>

  <dependencies>
    <dependency>
      <groupId>com.fasterxml.woodstox</groupId>
      <artifactId>woodstox-core</artifactId>
      <version>5.0.1</version>
    </dependency>
    <dependency>
      <groupId>xerces</groupId>
      <artifactId>xercesImpl</artifactId>
      <version>2.11.0</version>
    </dependency>
    <dependency>
      <groupId>com.fasterxml</groupId>
      <artifactId>aalto-xml</artifactId>
      <version>1.0.0</version>
    </dependency>
  </dependencies>

  <build>
    <plugins>
      <plugin>
        <artifactId>maven-compiler-plugin</artifactId>
        <version>3.3</version>
        <configuration>
          <source>1.8</source>
          <target>1.8</target>
        </configuration>
      </plugin>
      </plugins>
  </build>
</project>
```

Output:

```
************** Trying AALTO **************
-------------- Factory classes:
XMLInputFactory=com.fasterxml.aalto.stax.InputFactoryImpl
XMLOutputFactory=com.fasterxml.aalto.stax.OutputFactoryImpl
XMLEventFactory=com.fasterxml.aalto.stax.EventFactoryImpl
-------------- StAX results:
Buffered text: [a
a]
Code point at index 1: 13
Buffered text equals input text? false
Buffered text equals expanded text? false
Buffered text has \n for \r? false
StAX XML: [<x>aa</x>]
============== DOM results:
Buffered text: [aa]
Code point at index 1: 97
Buffered text equals input text? false
Buffered text equals expanded text? false
Buffered text has \n for \r? false

************** Trying JDK **************
-------------- Factory classes:
XMLInputFactory=com.sun.xml.internal.stream.XMLInputFactoryImpl
XMLOutputFactory=com.sun.xml.internal.stream.XMLOutputFactoryImpl
XMLEventFactory=com.sun.xml.internal.stream.events.XMLEventFactoryImpl
-------------- StAX results:
Buffered text: [a
a]
Code point at index 1: 13
Buffered text equals input text? false
Buffered text equals expanded text? false
Buffered text has \n for \r? false
StAX XML: [<x>a
a</x>]
============== DOM results:
Buffered text: [a
a]
Code point at index 1: 10
Buffered text equals input text? false
Buffered text equals expanded text? false
Buffered text has \n for \r? true

************** Trying WOODSTOX **************
-------------- Factory classes:
XMLInputFactory=com.ctc.wstx.stax.WstxInputFactory
XMLOutputFactory=com.ctc.wstx.stax.WstxOutputFactory
XMLEventFactory=com.ctc.wstx.stax.WstxEventFactory
-------------- StAX results:
Buffered text: [a
a]
Code point at index 1: 13
Buffered text equals input text? false
Buffered text equals expanded text? false
Buffered text has \n for \r? false
StAX XML: [<x>a&#xd;a</x>]
============== DOM results:
Buffered text: [a
a]
Code point at index 1: 13
Buffered text equals input text? false
Buffered text equals expanded text? false
Buffered text has \n for \r? false

************** Trying XERCES **************
-------------- Factory classes:
XMLInputFactory=com.sun.xml.internal.stream.XMLInputFactoryImpl
XMLOutputFactory=com.sun.xml.internal.stream.XMLOutputFactoryImpl
XMLEventFactory=org.apache.xerces.stax.XMLEventFactoryImpl
-------------- StAX results:
Buffered text: [a
a]
Code point at index 1: 13
Buffered text equals input text? false
Buffered text equals expanded text? false
Buffered text has \n for \r? false
StAX XML: [<x>a
a</x>]
============== DOM results:
Buffered text: [a
a]
Code point at index 1: 10
Buffered text equals input text? false
Buffered text equals expanded text? false
Buffered text has \n for \r? true
```

**Comment by cowtowncoder** [comment]
Hmmh. One thing on test I do not understand is this:

    static final String CR = new String(Character.toChars(32));

which is NOT CR character but simple space.

**Comment by cowtowncoder** [comment]
I can reproduce this, should be possible to fix.

**Comment by conleym** [comment]
Terrible variable naming on `CR`, yeah...oops! Sorry for the confusion.

**Comment by cowtowncoder** [comment]
No problem, figured it out. Fix will be in 1.1.0, hope to release soon (this week ideally).
